### DISCUSSION TITLE
Getting Started With Java Development in 2023 — An Opinionated Guide - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#306] [Announcements]
# Getting Started With Java Development in 2023 — An Opinionated Guide - Gunnar Morling

27 years of age, and alive and kicking — The Java platform regularly comes out amongst the top contenders in rankings like the TIOBE index.
In my opinion, rightly so. The language is very actively maintained and constantly improved;
its underlying runtime, the Java Virtual Machine (JVM),
is one of, if not the most, advanced runtime environments for managed programming languages.

There is a massive eco-system of Java libraries which make it a great tool for a large number of use cases,
ranging from command-line and desktop applications, over web apps and backend web services, to datastores and stream processing platforms.
With upcoming features like support for vectorized computations (SIMD),
light-weight virtual threads,
improved integration with native code,
value objects and user-defined primitives, and others,
Java is becoming an excellent tool for solving a larger number of software development tasks than ever before.

https://www.morling.dev/blog/getting-started-with-java-development-2023/

**Comment by TylerHillery** [comment]
The vast Java community and ecosystem is definitely a strong point for Java but as a Java beginner it can make learning Java daunting so thank you for this guide!

**Reply by gunnarmorling** [reply]
Thank you so much, @TylerHillery. All the best for your Java journey!

**Comment by sormuras** [comment]
Nice guide, Gunnar!

~~Nit: I expected to find "Learning the Language" way more up in the list - right after "Your First Java Program".~~ Thanks for moving it up!

Tip: A more modern option to install a JDK on Windows is using [winget](https://github.com/microsoft/winget-cli).

- Run `winget search jdk` to list available JDK distributions.
- Run `winget install "Java SE Development Kit 17"` to install JDK 17.

**Reply by gunnarmorling** [reply]
Hey Christian, thanks for your feedback. Good point on winget, I've added a note on that (I wouldn't know, haven't touched Windows for 10+ years ;).

**Comment by rdeviprasad** [comment]
A really detailed and nice guide for beginners. Great work Gunnar!!

**Reply by gunnarmorling** [reply]
Thanks, @rdeviprasad!

**Comment by michael-simons** [comment]
Excellent post. 

I would probably mention that `java` will just happily run single source `.java` files without compiling them explicitly first.

**Reply by sormuras** [reply]
The single-file source-code program launch feature is mentioned in the 

💡 **Your First Java Program**

section.

**Reply by michael-simons** [reply]
Aah yes. I just looked for it around jshell and jbang

**Reply by michael-simons** [reply]
Tja, skimming the information I’m interested in, complaining what I think should be there and not seeing the obvious. I feel slightly embarrassed 😅

**Reply by gunnarmorling** [reply]
Hehe, no worries 😃. Thanks for reading and commenting, @michael-simons. Btw. @sormuras, I've pulled up "Learning the Language", you were not the only one with that remark. Thanks again.

**Comment by mattwelke** [comment]
This is the guide I wish I had when I was trying to wrap my head around Java. I like that you recommend a specific build tool, IDE, and a set of libraries for common tasks like JSON processing.

**Reply by gunnarmorling** [reply]
Hey Matt, thanks a lot for the nice feedback!

**Comment by corneil** [comment]
I would suggest listing GraalVM is a JVM implementation alongside Hotspot and OpenJ9.
